### PR TITLE
Mingw cross-compiling

### PR DESCRIPTION
**Pull request by toots**
Hi,

I've had some interesting time compiling a cross-compiler for mingw. 

I've used Debian testing and this configure command:

```
./configure -no-ocamldoc -no-ocamlbuild \
            -target x86_64-w64-mingw32 \
            -target-bindir ~/x86_64-w64-mingw32/bin
```

You need to compile a custom `flexdll` for `dynlink` to be enabled, otherwise disable with `-no-shared-libs`

Then:

```
make world.opt
```

Will create a `ocamlopt` that you can run with your host `ocamlrun` and which will produce valid 64 bit windows binary.

It also works for 32 bit windows but you'll have to pre-compile a 32 bit runtime.

This PR contains some of the changes, though I am not sure what should be merged as-is. Also, not sure about installation.. :-)
- Commit 5c8058e contains mostly warning fixes and should be fine to merge as-is.
- Commit 2257bd9 makes `runtest` return with exit code `100` when test program is successfully compiled but fails to run, which happens in this case
- Commit 2634e45 fix reporting of OS as `"Win32"`
- Commit 8f6025c fixes some issues with `printf` format using mingw/native windows API.
- Commit a5e7875 changes `SYSTEM` variable to `mingw64` for 64 bit mingw compilation.
- Commit 33683aa makes a couple of seemingly innocuous Makefile tweaks.

### CODE REVIEW

**Comment by toots** [comment]
Actually, it also works pretty well for 32 bit windows, provided you have a 32 bits `ocamlrun` in the `PATH`..

File: otherlibs/systhreads/Makefile [hunk 1]
```diff
@@ -19,9 +19,9 @@ CAMLYACC ?= ../../boot/ocamlyacc
 
 ROOTDIR=../..
 CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib \
-      -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix
+      -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix -I $(ROOTDIR)/otherlibs/win32unix
```
Comment (lijunsong) [inline otherlibs/systhreads/Makefile]: How about something along these lines to not include the path on other non-windows systems.

```
ifeq ($(TOOLCHAIN),mingw)
INCLUDE_WINPATH=-I $(ROOTDIR)/otherlibs/win32unix
else
INCLUDE_WINPATH=
endif
 ROOTDIR=../..
 CAMLC=$(CAMLRUN) $(ROOTDIR)/ocamlc -nostdlib \
      -I $(ROOTDIR)/stdlib -I $(ROOTDIR)/otherlibs/unix $(INCLUDE_WINPATH)
```

**Comment by toots** [comment]
Excellent idea @lijunsong, just pushed it.

**Comment by mshinwell** [comment]
@shindere This is probably in your neck of the woods as well.

**Comment by shindere** [comment]
Mark Shinwell (2016/12/28 02:54 -0800):
> @shindere This is probably in your neck of the woods as well.

Taken into account, thanks!

**Comment by XVilka** [comment]
Is this still relevant?

**Comment by toots** [comment]
Clearly not :-)

I'll close this one. We've been maintaining windows cross-compilers over at https://github.com/ocaml-cross/opam-cross-windows. Is there an effort to get this to work in the main OCaml code? If so, is there a person of contact to try to backport the last remaining patches to achieve that?

**Comment by shindere** [comment]
Hi,

Romain Beauxis (2019/09/20 08:29 -0700):
> Clearly not :-)
>
> I'll close this one. We've been maintaining windows cross-compilers
> over at https://github.com/ocaml-cross/opam-cross-windows. Is there an
> effort to get this to work in the main OCaml code?

Yes

> If so, is there a person of contact to try to backport the last remaining patches to
> achieve that?

Please, feel fre to contact me. Sebastien.Hinderer@inria.fr

**Pull request closed without merging**